### PR TITLE
feat: add new benchmark Stable Diffusion

### PR DESCRIPTION
**Pull request by anhappdev**
Relevant files can be downloaded from here:
https://drive.google.com/drive/folders/10zCF7_ctUIM7xVPyhw5P60utnvWVUcSy?usp=sharing

### CODE REVIEW

**Comment by anhappdev** [comment]
@RSMNYS You should be able to integrate the Stable Diffusion pipeline for mobile_back_tflite based on this branch.
I will implement the download of multiple models in a separated PR. But for now, you should be able to just use a local folder path instead of a file path, ([more here](https://github.com/mlcommons/mobile_app_open/blob/09e4b41f72714c84fb0cf844433da47ff726f62f/docs/custom-tasks.md#using-external-resources-on-an-android)) e.g.:
```pbtxt
model_path: "local:///path/to/folder"
```
then the model_path will be a folder path in:
```cpp
mlperf_backend_ptr_t mlperf_backend_create(const char* model_path, mlperf_backend_configuration_t* configs, const char* native_lib_path)
```

File: flutter/cpp/datasets/coco_gen.cc [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    // [-1.0, 1.0] -> [0, 255]
+    for (int i = 0; i < OUTPUT_SIZE; i++) {
+      output_pixels[i] = (uint8_t)((*(temp_data + i) + 1) / 2 * 255);
+    }
+  }
+  if (!output_pixels.empty()) {
+    CaptionRecord* record = samples_.at(sample_idx).get();
+    auto input_ids = record->get_input_ids_vector();
+    auto attention_mask = record->get_attention_mask_vector();
+    // TODO: resize output_pixels to 224x224 and assign to pixel_values
```
Comment (anhappdev) [inline flutter/cpp/datasets/coco_gen.cc:52]: @freedomtan Do you have any suggestion how we can resize the image here?
The `tflite::evaluation::ImagePreprocessingStage` can only read from file path, so I cannot use it here.

Comment (anhappdev) [reply flutter/cpp/datasets/coco_gen.cc]: I resolve this one by adding a resizing layer to the CLIP model. The notebook to convert the CLIP model to TFLite is included in the PR.

**Comment by freedomtan** [comment]
my old resize() with TFLite resize op (https://github.com/tensorflow/tensorflow/blob/2ca3e974e350664c0e18b2ff5a52867490a15222/tensorflow/lite/examples/label_image/bitmap_helpers_impl.h#L30)

**Comment by freedomtan** [comment]
@freedomtan to check this implementation

**Comment by RSMNYS** [comment]
> @RSMNYS You should be able to integrate the Stable Diffusion pipeline for mobile_back_tflite based on this branch. I will implement the download of multiple models in a separated PR. But for now, you should be able to just use a local folder path instead of a file path, ([more here](https://github.com/mlcommons/mobile_app_open/blob/09e4b41f72714c84fb0cf844433da47ff726f62f/docs/custom-tasks.md#using-external-resources-on-an-android)) e.g.:
> 
> ```textproto
> model_path: "local:///path/to/folder"
> ```
> 
> then the model_path will be a folder path in:
> 
> ```c++
> mlperf_backend_ptr_t mlperf_backend_create(const char* model_path, mlperf_backend_configuration_t* configs, const char* native_lib_path)
> ```

Hi guys! I've pushed the code partially here:

https://github.com/mlcommons/mobile_app_open/tree/features/pipeline_backend

The idea was to create the pipeline (stable_diffusion_backend.cc). And initiate it from the dart-run-benchmark.cc, by checking the dataset type for example. Then in issueQuery we can process the stable diffusion steps. 

@anhappdev @freedomtan

**Comment by anhappdev** [comment]
@RSMNYS I refactored the TFLite backend to support multiple pipelines. You can add your implementation to `mobile_back_tflite/cpp/backend_tflite/stable_diffusion_pipeline.cc`.

File: flutter/cpp/datasets/coco_gen.cc (hunk 1 above)
Comment (anhappdev) [reply flutter/cpp/datasets/coco_gen.cc]: > my old resize() with TFLite resize op (https://github.com/tensorflow/tensorflow/blob/2ca3e974e350664c0e18b2ff5a52867490a15222/tensorflow/lite/examples/label_image/bitmap_helpers_impl.h#L30)

In code, we will need to have 2 steps: resize image and convert data layout  from NHWC to NCHW. This is already integrated in the TFLite model.
After some consideration, I think it's better to have the steps integrated in the TFLite model.

**Review by freedomtan — DISMISSED** [review]
LGTM

File: flutter/cpp/datasets/coco_gen.cc [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+#include "coco_gen.h"
+
+#include <iomanip>
+
+namespace mlperf {
+namespace mobile {
+namespace {}  // namespace
+
+CocoGen::CocoGen(Backend* backend, const std::string& input_tfrecord,
```
Comment (mohitmundhragithub) [inline flutter/cpp/datasets/coco_gen.cc:21]: Hi @anhappdev , 

do we have the Accuracy dataset and the input_clip_model shared anywhere? can you please share the link?

Comment (anhappdev) [reply flutter/cpp/datasets/coco_gen.cc:21]: The files are stored in Google Drive linked in the first comment of this PR.

Comment (mohitmundhragithub) [reply flutter/cpp/datasets/coco_gen.cc:21]: @anhappdev, what should be the input format for the clip_model.tflite? is it required to be NHWC or NCHW?

Comment (anhappdev) [reply flutter/cpp/datasets/coco_gen.cc:21]: The `clip_model_512x512.tflite` is the one to be used by the app, and the input should be 512x512 image in NHWC layout.
You can see the conversion script at `flutter/cpp/datasets/coco_gen_utils/CLIP_Model_to_TFLite.ipynb`

Comment (mohitmundhragithub) [reply flutter/cpp/datasets/coco_gen.cc:21]: do we have the tfrecord file decided for the validation set?

Comment (anhappdev) [reply flutter/cpp/datasets/coco_gen.cc:21]: No, not yet. But you can easy generate it with the script at `flutter/cpp/datasets/coco_gen_utils/generate_tfrecords.py`

Related comment:
https://github.com/mlcommons/mobile_model_closed/issues/9#issuecomment-2319660564

Comment (anhappdev) [reply flutter/cpp/datasets/coco_gen.cc:21]: I uploaded a `coco_gen_test.tfrecord` with 150 prompts to the Google Drive folder.

**Comment by anhappdev** [comment]
We can merge this PR into master now. I disabled the tests for `stable_diffusion` task, so we have it passed the CI when merging into master and will enable it again in the submission branch.

**Review by freedomtan — APPROVED** [review]
LGTM. @mohitmundhragithub if you are OK with this. Let's merge it.
